### PR TITLE
Histograms

### PR DESCRIPTION
**Pull request by stijn-devriendt**
This changeset mainly adds support for histograms to the prometheus client library.

It also fixes a small bug wrt gettimeofday, improves compliance to guidelines (https://prometheus.io/docs/instrumenting/writing_clientlibs/) and allows reading back counters from code.

Histograms add extra labels (named "le") to the user-provided set. This support was lacking in the original version and was added. It can be reused for adding quantile support to the summary counters.

### CODE REVIEW

**Comment by stijn-devriendt** [comment]
Array.create_float was intoduced in 4.03, leaving me with a couple of options:
- use make_float (which is deprecated and might yield deprecation warnings)
- introduce cppo. A bit too heavy, I guess?
- use Array.make, not using the benefits of make_float.
- bump the minimum required compiler version

Any input/preference would be appreciated.

**Comment by hannesm** [comment]
@stijn-devriendt with the pace of OCaml releases, and MirageOS move towards 4.04.2 only (on Xen) and >=4.03 on kvm+virtio, I'd guess it is fine to drop <4.03 support in this library.

File: src/prometheus.ml [hunk 1]
```diff
@@ -236,8 +251,150 @@ module Summary = struct
...
+  }
+
+let _histogram at_index_f count =
+  let real_at_index i =
+    if i >= count then
+      infinity
+    else
+      at_index_f i
+  in
+  let upper_bounds = (Array.create_float (count + 1)) in
```
Comment (talex5) [inline src/prometheus.ml]: I think you can just use `Array.init` here and avoid a lot of complexity, and continue to work with all OCaml versions.

File: src/prometheus.ml [hunk 2]
```diff
@@ -236,8 +251,150 @@ module Summary = struct
...
       )
 end
+
+type histogram = {
+    upper_bounds: float array;
+    counts: float array;
+    mutable sum: float;
+  }
+
+let _histogram at_index_f count =
```
Comment (talex5) [inline src/prometheus.ml]: Why does this start with an underscore?

File: src/prometheus.ml [hunk 3]
```diff
@@ -81,11 +79,11 @@ module MetricFamilyMap = Map.Make(MetricInfo)
 
 module CollectorRegistry = struct
   type t = {
-    mutable metrics : (unit -> (string * float) list LabelSetMap.t) MetricFamilyMap.t;
+    mutable metrics : (unit -> (string * float * ((LabelName.t * float) option)) list LabelSetMap.t) MetricFamilyMap.t;
     mutable pre_collect : (unit -> unit) list;
   }
 
-  type snapshot = (string * float) list LabelSetMap.t MetricFamilyMap.t
+  type snapshot = (string * float * ((LabelName.t * float) option)) list LabelSetMap.t MetricFamilyMap.t
```
Comment (talex5) [inline src/prometheus.ml:86]: Why do we need a third component to the tuple here? It looks like we have a list, plus an optional extra element to add to the list, which seems pointless.

File: src/prometheus.mli [hunk 4]
```diff
@@ -113,6 +114,9 @@ module Counter : sig
   val inc_one : t -> unit
   val inc : t -> float -> unit
   (** [inc t v] increases [t] by [v], which must be non-negative. *)
+
+  val get : t -> float
+  (** [get t] returns the current value of the counter. *)
```
Comment (talex5) [inline src/prometheus.mli:119]: Why are we exposing this?

File: src/prometheus.ml (hunk 3 above)
Comment (stijn-devriendt) [reply src/prometheus.ml:86]: The string bit of the list contains a suffix for the metric name, the added parameter is for a label.
i.e.:
["", 3.0, Some ("blah", 3.0)] ->  "metricname{blah="3.0"} 3.0"
["_sum", 3.0, None] -> "metricname_sum 3.0"

File: src/prometheus.mli (hunk 4 above)
Comment (stijn-devriendt) [reply src/prometheus.mli:119]: In our use-case we'd like to use the accumulated values for internal decision-making and posting some of the values over another interface towards another system. To do that we want to read back the embedded values.

**Comment by stijn-devriendt** [comment]
I've replaced the ocaml compiler bump commit with the reworked array initialization.

**Comment by talex5** [comment]
(this was mostly merged in #14 - feel free to open a new PR if you feel that didn't add all the features you need)

**Pull request closed without merging**